PlaceRouteHierFlow/placer/ILP_solver.cpp

```diff
@@ -1130,20 +1127,29 @@ double ILP_solver::GenerateValidSolution(design& mydesign, SeqPair& curr_sp, PnR
     }
   }
 
-  // x>=0, y>=0
-  for (auto id : curr_sp.negPair) {
-    if (id < int(mydesign.Blocks.size())) {
-      // x>=0
-      {
-        double sparserow[1] = {1};
-        int colno[1] = {id * 4 + 1};
-        if (!add_constraintex(lp, 1, sparserow, colno, GE, 0)) logger->error("error");
+
+  if (flushbl) {
+    for (auto id : curr_sp.negPair) {
+      if (id < int(mydesign.Blocks.size())) {
+        // x>=0
+        set_bounds(lp, (id * 4 + 1), 0, get_infinite(lp));
+        // y>=0
+        set_bounds(lp, (id * 4 + 2), 0, get_infinite(lp));
       }
-      // y>=0
-      {
-        double sparserow[1] = {1};
-        int colno[1] = {id * 4 + 2};
-        if (!add_constraintex(lp, 1, sparserow, colno, GE, 0)) logger->error("error");
+    }
+  } else {
+    // x>=0, y>=0
+    int minx{0}, miny{0};
+    for (auto id : curr_sp.negPair) {
+      if (id < int(mydesign.Blocks.size())) {
+        minx += mydesign.Blocks[id][curr_sp.selected[id]].width;
+        miny += mydesign.Blocks[id][curr_sp.selected[id]].height;
+      }
+    }
+    for (auto id : curr_sp.negPair) {
+      if (id < int(mydesign.Blocks.size())) {
+        set_bounds(lp, (id * 4 + 1), -10*minx, -mydesign.Blocks[id][curr_sp.selected[id]].width);
+        set_bounds(lp, (id * 4 + 2), -10*miny, -mydesign.Blocks[id][curr_sp.selected[id]].height);
```

Comment:
Why the bound is negative?

Reply:
Using the following convention:
flush left: left edge flushes to "0".
![image](https://user-images.githubusercontent.com/73570678/140977187-063a9978-625c-45a5-a2b2-31d3b7e4d4ad.png)

flush right: right edge flushes to "0".
![image](https://user-images.githubusercontent.com/73570678/140977066-546505a0-f686-4022-bb3a-3f5d25abfd85.png)

This makes the coordinates negative. They are shifted back to origin in lines 1562:1565.

Reply:
I see. Thanks.